### PR TITLE
feat: add specifications for new operators

### PR DESCRIPTION
**Pull request by ivarconr**

### CODE REVIEW

File: schema/features-schema.js [hunk 1]
```diff
@@ -17,7 +17,10 @@ const schema = Joi.object().keys({
                         Joi.object().keys({
                             contextName: Joi.string().required(),
                             operator: Joi.string().required(),
-                            values: Joi.array().items(Joi.string())
+                            values: Joi.array().items(Joi.string()),
+                            caseInsensitive: Joi.bool().optional(),
+                            inverted: Joi.bool().optional(),
+                            value: Joi.alternatives().try(Joi.string(), Joi.number()).optional()
```
Comment (ivarconr) [inline schema/features-schema.js:23]: @chriswk will it cause problems in java if we allow the value property to be either a number or a string?

File: specifications/13-constraint-operators.json [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                "enabled": true,
+                "strategies": [
+                    {
+                        "name": "default",
+                        "parameters": {},
+                        "constraints": [
+                            {
+                                "contextName": "someValue",
+                                "operator": "NUM_EQ",
+                                "value": 12
```
Comment (ivarconr) [inline specifications/13-constraint-operators.json:132]: We might consider to use string values only, and have the SDK parse it to a number type.

**Comment by ivarconr** [comment]
The specifications runs green in the node SDK implementation of them 🔥 ([PR](https://github.com/Unleash/unleash-client-node/pull/289))

File: schema/features-schema.js (hunk 1 above)
Comment (chriswk) [reply schema/features-schema.js:23]: Sorry for the delayed answer, but yes, I don't think Gson knows how to handle a type incertainity. This will require a major version, because it is a backwards incompatible change to go from a guaranteed string to either/or

File: specifications/13-constraint-operators.json (hunk 2 above)
Comment (chriswk) [reply specifications/13-constraint-operators.json:132]: I think we're going to have to if we don't want to break the api contract with earlier sdks, but that from the operator, the sdk's will have to do the parsing.

**Review by chriswk — CHANGES_REQUESTED** [review]
value/values needs to stay a string

File: schema/features-schema.js (hunk 1 above)
Comment (ivarconr) [reply schema/features-schema.js:23]: make sense. We could have seperate fields per type, aka `dateValue`, `strValue` and so on, but feels a bit verbose. I vote for keeping it as a string as well.

File: specifications/13-constraint-operators.json (hunk 2 above)
Comment (ivarconr) [reply specifications/13-constraint-operators.json:132]: lets keep it as a string.

File: schema/features-schema.js (hunk 1 above)
Comment (RikudouSage) [reply schema/features-schema.js:23]: I agree with keeping it a string, could be changed to number in next major, seems pointless to do a major release due to such small change. The SDKs could parse the string to correct format easily based on the operator.

The `dateValue`, `strValue` etc. indeed does sound too verbose.

File: specifications/13-constraint-operators.json [hunk 3]
```diff
@@ -0,0 +1,729 @@
...
+                "enabled": true,
+                "strategies": [
+                    {
+                        "name": "default",
+                        "parameters": {},
+                        "constraints": [
+                            {
+                                "contextName": "currentTime",
+                                "operator": "DATE_AFTER",
+                                "value": "2022-01-29T13:00:00.000Z"
```
Comment (ivanklee86) [inline specifications/13-constraint-operators.json:264]: This is probably a minor thing but!  Should `DATE_BEFORE` and `DATE_AFTER` be inclusive (i.e. LT versus LTE)? 

I don't see this being a huge issue (there's a vanishingly small chance people will notice) but figured I'd ask as long as I was writing unit tests.  In the Python client, it's currently coded to be inclusive.

**Comment by RikudouSage** [comment]
I'm missing some edge cases for versions, like 1.2.3-beta being lower than 1.2.3 and greater than 1.2.3-alpha.

**Comment by chriswk** [comment]
Good catch @RikudouSage - they are indeed missing, and we should add those as well :)

File: schema/features-schema.js (hunk 1 above)
Comment (ivarconr) [reply schema/features-schema.js]: We use just `value` when the operator only uses a single value, and it will always be a string, requiring the SDK to parse it per operator.

File: specifications/13-constraint-operators.json (hunk 3 above)
Comment (ivarconr) [reply specifications/13-constraint-operators.json:264]: The meaning of the word "AFTER" seems to be closer to GT. 

For this reason I think we should require it to be exclusive.

File: schema/features-schema.js (hunk 1 above)
Comment (ivarconr) [reply schema/features-schema.js]: fixed!

**Comment by ivarconr** [comment]
I think we are close to a an agreeable state on the specs. 

Would be good to get feedback from 
- @ivarconr (node.sdk and proxy)
- @ivanklee86 (python SDK)
- @RikudouSage (php SDK 
- @chriswk (java SDK)
- @rarruda (ruby sdk)
- @FredrikOseberg (frontend)
- @thomasheartman (docs)

**Review by ivarconr — COMMENTED** [review]
lgtm

**Review by chriswk — APPROVED** [review]
This looks good now :)

**Comment by ivanklee86** [comment]
Updated spec tests in  https://github.com/Unleash/unleash-client-python/pull/192 and everything looks good. :)

**Review by RikudouSage — APPROVED** [review]
Looks good to me, updated to latest version in https://github.com/Unleash/unleash-client-php/pull/47 and everything works.

File: specifications/13-constraint-operators.json (hunk 3 above)
Comment (rarruda) [reply specifications/13-constraint-operators.json:282]: * How about also allowing using `DATE_GTE`, `DATE_GT` as more precise definitions of `DATE_AFTER`?
  (and the equivalent for `_LTE`, `_LE` of for `DATE_BEFORE`)

* It would have been useful to also have `SEMVER_GTE` and `_LTE`

* It would also nice to have tests for other strings which are definitely not SEMVER, so we have predictable behaviour too. Maybe just do in the current locale string ordering when one of the values is not a SEMVER? (optionally log a warning?) Like :
  * "v1.0.0" <=> "1.0.0"  ( I know that [versions prefixed with `v` are not SEMVER](https://semver.org/#is-v123-a-semantic-version) but they are still widely use, so it would be good that we at least test for them)
  * "2.0.0-beta.1" <=> "2.0.0b1"
  * "abc" <=> "abz"
  * "1.0" <=> "abc"

I understand that this feedback might be a bit late, as the spec seems to be closed now, but I thought it was still worth giving my 2 cents.

**Comment by rarruda** [comment]
* How about also allowing using `DATE_GTE`, `DATE_GT` as more precise definitions of `DATE_AFTER`?
  (and the equivalent for `_LTE`, `_LE` of for `DATE_BEFORE`)

* It would have been useful to also have `SEMVER_GTE` and `_LTE`

* It would also nice to have tests for other strings which are definitely not SEMVER, so we have predictable behaviour too. Maybe just do in the current locale string ordering when one of the values is not a SEMVER? (optionally log a warning?) Like :
  * "v1.0.0" <=> "1.0.0"  ( I know that [versions prefixed with `v` are not SEMVER](https://semver.org/#is-v123-a-semantic-version) but they are still widely use, so it would be good that we at least test for them)
  * "2.0.0-beta.1" <=> "2.0.0b1"
  * "abc" <=> "abz"
  * "1.0" <=> "abc"

I understand that this feedback might be a bit late, as the spec seems to be closed now, but I thought it was still worth giving my 2 cents.

**Comment by RikudouSage** [comment]
> * How about also allowing using `DATE_GTE`, `DATE_GT` as more precise definitions of `DATE_AFTER`?
>       (and the equivalent for `_LTE`, `_LE` of for `DATE_BEFORE`)

This doesn't make much sense IMO, since the current time would need to be the same down to exactly seconds (or milliseconds, microseconds, whatever your implementation takes into consideration).

**Comment by rarruda** [comment]
@RikudouSage Given that `DATE_AFTER` is strictly after (exclusive), we kind of are in a conundrum anyhow:

How do you represent the from 1pm and on for a given date?

using `DATE_AFTER` with "2022-01-29T13:00:00.000Z" then:
* "2022-01-29T13:00:00.000Z" will be excluded
* "2022-01-29T13:00:00.001Z" will be included.

If you want to include "2022-01-29T13:00:00.000Z" as well, then you could set `DATE_AFTER` to "2022-01-29T12:59:59.999Z", but what if there is an event at "2022-01-29T12:59:59.9999Z" that you don't want included?

**Comment by rarruda** [comment]
I understand your concern regarding "exactly" equals time, but It should be safe to assume that non defined precisions are floored down to zero.

**Comment by ivarconr** [comment]
> It would also nice to have tests for other strings which are definitely not SEMVER, so we have predictable behaviour too.

We constantly should add more edge cases as we discover them. As long as the correct behavior is possible to document and within the agreed protocol any deviation should be considered as a bug in the respective SDKs.

For DATE operators I do not worry much about micro-precision. If you have the need to be precise on this level you probably also worry about time-synchronization between servers and would probably not rely on any third part tool for your business logic. 

> It would have been useful to also have SEMVER_GTE and _LTE

Maybe, we can add these later. 

> How about also allowing using DATE_GTE, DATE_GT as more precise definitions of DATE_AFTER?
(and the equivalent for _LTE, _LE of for DATE_BEFORE)

These date operators where on purpose given a different name, just because they are "less accurate". The chances of a date-time to be equal on a sub-second level is quite small, and only adds additional options the user needs to consider.
